### PR TITLE
[2201.4.0-stage] Fix no such type error when anonymous error types in imported modules are used

### LINKED ISSUE #39453: [Bug]: Getting No such type error when building graphql test package
### Description

Error encountered while executing the command `gradle test`:
```
error: No such type: Package: ballerina/graphql/1/$_init,
          TypeName: $type$$anonType$_17, Shape: record {| ballerina/graphql:1.6.0:ErrorDetail[]? errors; |}
```
The graphql repo has two modules:
1. Main module (`ballerina`)
2. Test module (`ballerina-test`)

the main module includes a submodule called `parser` which defines the following type: [type](https://github.com/ballerina-platform/module-ballerina-graphql/blob/master/ballerina/modules/parser/records.bal#L38)
```ballerina
public type ErrorDetail record {|
    string message;
    Location[] locations?;
    (int|string)[] path?;
    map<anydata> extensions?;
|};
```
The main module also defines the following [type](https://github.com/ballerina-platform/module-ballerina-graphql/blob/master/ballerina/records.bal#L135) and [errors](https://github.com/ballerina-platform/module-ballerina-graphql/blob/master/ballerina/errors.bal#L47)

```ballerina
public type ErrorDetail record {|
    *parser:ErrorDetail;
|}; 

public type InvalidDocumentError distinct (RequestError & error<record {| ErrorDetail[]? errors; |}>);
public type PayloadBindingError distinct (ClientError & error<record {| ErrorDetail[]? errors; |}>);
```

Note that when the command `gradle clean build` is executed, both main and test modules' tests run without any issue. But when the command `gradle test` is executed, tests from the main module run as expected but tests from the test module fail with the above mentioned error.

Note: this [test](https://github.com/ballerina-platform/module-ballerina-graphql/blob/master/ballerina-tests/tests/35_client_test.bal#L175) in the test module uses one of the aforementioned error types.

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
Fixes #39453

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by HindujaB — APPROVED** [review]
LGTM

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/public_and_private_types_test.bal [hunk 1]
```diff
@@ -24,3 +24,9 @@ function testModulePublicAndPrivateTypes() {
         test:assertEquals(timeZone.getInt({year:1, month: 2, day: 3}), 6);
     }
 }
+
+function testAnonymousDistinctErrorTypes() {
+    error err = error("");
+    boolean ans = err is test_module:InvalidDocumentError;
+    test_module:InvalidDocumentError|error err2 = trap <test_module:InvalidDocumentError>err;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/public_and_private_types_test.bal:31]: Can we add an assertion also? Maybe a positive and negative case that will help assert that the type is correct also.

Comment (Nadeeshan96) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/public_and_private_types_test.bal:31]: Will do this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeGen.java [hunk 2]
```diff
@@ -698,7 +698,8 @@ private void loadErrorType(MethodVisitor mv, BErrorType errorType) {
             return;
         }
 
-        if (Symbols.isFlagOn(errorType.flags, Flags.ANONYMOUS)) {
+        boolean samePackage = JvmCodeGenUtil.isSameModule(this.packageID, pkgID);
+        if (Symbols.isFlagOn(errorType.flags, Flags.ANONYMOUS) && samePackage) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeGen.java:702]: IMO correct fix would be to fix the hashcode and call get anon type. Shall we create an issue for that

Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeGen.java:702]: Created https://github.com/ballerina-platform/ballerina-lang/issues/39483.

**Review by pcnfernando — APPROVED** [review]

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**